### DISCUSSION TITLE
Question regarding `--import-mode=importlib`

### DISCUSSION THREAD

**Discussion by Kludex** [pytest-dev/pytest#12714] [Q&A] [answered]
I have this project layout:

```
src/
   \-> a-a
        \-> a_a
        \-> tests/
   \-> b-b
        \-> b_b
        \-> tests/
```

On `b-b/tests` I have a utility module called `utils`. Which means I import it with `tests.utils`.

As far as I understand, `--import-mode=importlib` doesn't add `tests` to the `PYTHONPATH`, but I have the `ModuleNotFound` error for `tests.utils`, and when I debug, what I see is that the `tests` imported is actually the one in `a-a/tests`.

So... I'm a bit confused why the first `tests` folder that `pytest` finds is added to the PYTHONPATH, or is this a Python thing?

If I add `testspath`, and list the tests folders it works as expected.

**Comment by nicoddemus** [comment]
Hi @Kludex 

Is that reproducible in a MWE? I ask because some things are omitted from the post, for example if there are `__init__.py` files, or how do you execute pytest (`pytest` vs `python -m pytest`), what is the `rootdir`, etc.

**Reply by nicoddemus** [reply]
One thing pytest does is to attempt to make a module name from the module file path, usually relative to the rootdir:

https://github.com/pytest-dev/pytest/blob/38ad84bafd18d15ceff1960d636c693560337844/src/_pytest/pathlib.py#L695-L723

**Reply by Kludex** [reply]
Here: https://github.com/Kludex/jik/pull/1 🙏 

Thanks.

**Reply by nicoddemus** [reply]
@Kludex that is not really a Minimal Working Example (emphasis on "minimal" hehehe). But thanks anyway for taking the time to prepare it.

I did try to set it up but unfortunately it does not work on Windows (which I use for my main development):

```
  ChefBuildError

  Backend subprocess exited when trying to invoke get_requires_for_build_wheel
  ...
  RuntimeError: uvloop does not support Windows at the moment
```

But looking at the repo I see what I think is the cause: `jik-app` and `jik-server` each have a `pyproject.toml`, which is of course fine, but is what pytest uses to derive a `rootdir`. The `rootdir`  is then used to produce the module name from the module path by the `module_name_from_path` function that I linked above.

This is the intended behavior.

Given `jik-app` and `jik-server` are separate projects, executing pytest from `src` might result in namespace clashes, because collecting from both at the same time will result in a single `tests` namespace being created.

pytest currently does not support running test suites from multiple projects at the same time, given pytest only has the concept of a single `rootdir` (which is to say that it might work depending on the project).

In your case I think the solution is to run `pytest` twice, once on each subdirectory `jik-app` and `jik-server`. Alternatively, you can rename each `tests` directory to a unique name, say `tests-app` and `tests-server`, but this might come with its own problems, not sure.

HTH

**Reply by Kludex** [reply]
> The rootdir is then used to produce the module name from the module path by the module_name_from_path function that I linked above.

Why is the `.` not the `rootdir` here? I also have a `pyproject.toml` there, and a `tests` directory.

**Reply by nicoddemus** [reply]
What `rootdir` pytest resolves to? Can you post the pytest header you get when running the tests (the one that shows `rootdir: ...`)?

**Reply by Kludex** [reply]
```
======================================================================= test session starts ========================================================================
platform darwin -- Python 3.12.0, pytest-8.3.2, pluggy-1.5.0
rootdir: /Users/marcelotryle/dev/personal/jik
configfile: pyproject.toml
plugins: anyio-4.4.0, mock-3.14.0
collecting ... 
```
So... It resolves to the root directory. 🤔

**Reply by nicoddemus** [reply]
Hmm I that's strange then. Guess I would need to debug that in order to see what is really happening. Any chance of hacking the code so it runs on Windows?

**Reply by Kludex** [reply]
Yep, give me a sec. Sorry for that.

**Reply by Kludex** [reply]
Done: https://github.com/Kludex/jik/pull/1/

(I think uvloop is the only issue for Windows...)

**Reply by nicoddemus** [reply]
Thanks, I can run the tests now, but I will have to continue this investigation later. Appreciate the time with the reproducer. 👍

**Reply by Kludex** [reply]
No worries. I was able to make it work with the workaround in the description anyway. I'm just really wondering why my first attempt didn't work as I'd expect.

Thanks Bruno.

**Reply by nicoddemus** [reply, marked as answer]
Hi @Kludex finally found the time to investigate a bit more.

Actually pytest will use [`resolve_pkg_root_and_module_name`](https://github.com/pytest-dev/pytest/blob/03e54712dd123138aa4c1035a027e17d0dd9a86c/src/_pytest/pathlib.py#L748) to derive the full test module name, so it will derive the package root based on the presence of `__init__.py` files on the file's directory and upwards. 

In your example, for `/projects/jik/src/jik-app/tests/test_app.py`, it finds `/projects/jik/src/jik-app/tests` as the package root, and then `/projects/jik/src/jik-app/tests` gets added as `tests` to `sys.modules`, which results in this clashing later when importing from `src/uvicorn`, as `tests.utils` will not refer to `src/uvicorn/tests/utils.py`, but to `src/jik-app/tests/utils.py` (which does not exist) due to `tests` already being a module in `sys.modules`.

So this is indeed the cause:

> Given jik-app and jik-server are separate projects, executing pytest from src might result in namespace clashes, because collecting from both at the same time will result in a single tests namespace being created.

Curiously, on `main` the test suite *works* but by accident because [`pythonpath` is configured in `pyproject.toml`](https://github.com/Kludex/jik/blob/2e67aa60c92223399b5878806118d6cecd563572/pyproject.toml#L31), which adds `src/uvicorn` to PYTHONPATH and causes `uvicorn/tests` to be import earlier... but this only masks the problem because a namespace clash might still be introduced in the future, in case someone adds some kind of testing utility module (say `foo`) to `jik-app/tests` or `jik-server/tests` and tries to import from there via `tests.foo` -- as `tests` in `sys.modules` will still be referencing `uvicorn/tests`.

**Reply by Kludex** [reply]
Ah, I see. Thanks Bruno. 🙏
